README.md
# meta-schema

[![Clojars Project](https://img.shields.io/clojars/v/meta-schema.svg)](https://clojars.org/meta-schema)


Designed to help data processing using `clojure.spec`.

Provide users the ability to combine several pre-defined
`clojure.spec` into some specific shape at *runtime*.


For now, it only supports JSON files. This library is in
alpha-stage, still needs many test cases and refactors,
although the Public API **must** not change.

## Installation

Leiningen/Boot

```clj
[meta-schema "0.1.4"]
```

Clojure CLI/deps.edn
```clj
meta-schema {:mvn/version "0.1.4"}
```

## Usage

In your project you need to define EDN files inside a folder
and provide a list of `java.io.File`s to the `(setup!  ..)`
function. These files are simple and provides the location
and intent of your pre-defined specs.

```clj
{:zipcode {:intent "Validation to ZipCodes in Brazil"
           :location :my-project.spec/zipcode-br}
 :money {:intent "How we understand money around here? Anything >= 0 either long, float, double, or decimal"
         :location :my-project.other-spec/money-br}
}
```

Make sure that all this namespaces are loaded before invoke
`meta-schema.core/setup!`.


Example of specification provided to the library at runtime.

```clj
  (require '[meta-schema.core :as ms])
  (require '[clojure.alpha.spec :as s])
  (require '[clojure.java.io :as io])

  (ms/setup! (-> (io/resources "specs")
                 (io/file)
                 (file-seq)))

  (def file-spec {:spec-name :my-project.client/payload
                  :zip [{:spec :zipcode
                         :optional? false
                         :nullable? false}

                        {:house-price {:spec :money
                                       :nullable? true}}]

                  :rent {:spec :money
                         :optional? false
                         :nullable? false}

                  :university {:departments [{:zip {:address {:spec :zipcode
                                                              :optional? false}}}]}})

  (def my-parser (ms/create-parser file-spec))

  (def my-data-1 {:zip ["my zipcode is text"]
                  :rent 10
                  :university {:departments [{:zip {:address "deep structures"}}]}})

  (s/valid? my-parser my-data-1) ;; => true

  (s/valid? my-parser {:zip [{:house-price 230} "either or both here.."]
                       :rent 10
                       :university {:departments [{:zip {:address "deep structures"}}]}}) ;; => true

```

Cool, now we can finally finish the example above. Let's say your source code only understands a specific `target-fmt` of MAP that can be processed. For sure, your clients does not want to send you that format, right? Because they can't change legacy systems and make something better to help everybody. That's ok!

```clj
  (def client-data {:zip ["101030-201", "987621-281"]
                    :rent 980.322
                    :university {:departments [{:zip {:address "University at Medium Inc.,"}}]}})

  (def client-spec {:spec-name :my-project.client/payload
                    :zip [{:spec :zipcode
                           :destination :zipcode}]

                    :rent {:spec :money
                           :destination :value}

                    :university {:departments [{:zip {:address {:spec :zipcode
                                                                :destination :university-address}}}]}})

  (def target-fmt {:my-internal-zipcode :zipcode
                   :my-internal-value   :value
                   :my-internal-address :university-address})

  (setup! (-> (io/resource "specs")
              (io/file)
              (file-seq)))

  (input-data->target-data client-data client-spec target-fmt)

  ;; => {:my-internal-zipcode ["101030-201" "987621-281"],
  ;;     :my-internal-value 980.322,
  ;;     :my-internal-address "University at Medium Inc.,"}
```

The clojure.spec to validate the data is created at runtime
and applied to the provided data. If it passes, we start the
transformation to the `target-shape` specified.

## Rationale

The DSL is simple:

1. When the field is a list, more than one spec inside it will produce a `or` operator
2. When the field is `optional`, it means that  may or may not be present at coersion time
3. When the field is `nullable`, it means that `nil` is a valid input.
4. By default all the fields are *not* nullable and *not* optionals.
5. You can nest the options as you desired


## TODO list

- [x] Finish the second part of the library, `input-data->target-data` conversions
- [x] Improve interfaces of public API
- [ ] Write a parser to spec the DSL in the input-file
- [ ] Provide support for `sets` and `enums` (?)
- [ ] Extend it to XML

## License

Copyright © 2020 Wanderson Ferreira (@bartuka)

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.1.4] - 2020-01-26
### Changed
- That was a mistake in the deployment to clojars with the
  public signature of a function. `input-data->target-data`
  was corrected and must remain the same now.

## [0.1.3] - 2020-01-26
### Changed
- Fixing typos in documentation


## [0.1.2] - 2020-01-26
### Added
- `input-fmt->target-fmt` to provide conversion between specs at runtime.


## 0.1.1 - 2020-01-20
- `create-parser` functional and public API defined.


test/meta_schema/core_test.clj
(ns meta-schema.core-test
  (:require [clojure.test :refer :all]
            [meta-schema.core :as ms]
            [clojure.java.io :as io]
            [clojure.spec.alpha :as s]
            [clojure.test.check.generators :as gen]
            [clojure.test.check.clojure-test :refer [defspec]]
            [clojure.test.check.properties :as prop]))

(def spec-gen (gen/hash-map
               :spec (gen/elements [:zipcode :money :life :address])))

(def optional-gen (gen/hash-map
                   :optional? gen/boolean))

(def nullable-gen (gen/hash-map
                   :nullable? gen/boolean))


(def leaf-spec (gen/let [specs spec-gen
                         options (gen/frequency [[6 optional-gen]
                                                 [4 (gen/return {})]])
                         nulls (gen/frequency [[6 nullable-gen]
                                               [4 (gen/return {})]])]
                 (merge specs options nulls)))

(def spec-file-gen (gen/hash-map
                    :spec-name gen/keyword-ns
                    :tk1 (gen/not-empty (gen/vector leaf-spec))
                    :tk2 leaf-spec))

(gen/sample spec-file-gen 2)


(s/def ::money (s/or :int int?
                     :float float?
                     :double double?))

(s/def ::zipcode string?)
(s/def ::life (s/and int? #(= 42 %)))
(s/def ::address string?)

(deftest simple-file-spec
  (ms/setup! (-> (io/resource "specs")
                 (io/file)
                 (file-seq)))
  (let [file-spec {:spec-name :meta-schema.core-test/testing
                   :cep {:spec :zipcode}
                   :amounts [{:spec :money}]}

        parser (ms/create-parser file-spec)]
    (is (s/valid? parser {:cep "brazilian cep"
                          :amounts [20 320 40 50]}))
    (is (not (s/valid? parser {:cep 10
                               :amounts 20})))))

(deftest failing-specs
  (ms/setup! (-> (io/resource "specs")
                 (io/file)
                 (file-seq)))
  (let [file-spec {:spec-name :meta-schema.core-test/test-fail
                   :cep {:spec :zipcode}
                   :address {:cep {:spec :zipcode}}
                   :money {:spec :money}}
        parser (ms/create-parser file-spec)]

    (is (not (s/valid? parser {:cep "br"
                               :address {:cep 10.0}
                               :money 20})))))


(defspec cannot-break-to-nested-fns 400
  (prop/for-all [file-spec spec-file-gen]
                (ms/setup! (-> (io/resource "specs")
                               (io/file)
                               (file-seq)))
                (= (spec-tools.core/spec? (ms/create-parser file-spec)))))


src/meta_schema/core.clj
(ns meta-schema.core
  (:require [clojure.edn :as edn]
            [clojure.string :as cstr]
            [spec-tools.data-spec :as ds]
            [meta-schema.dsl :refer :all :as dsl]
            [clojure.spec.alpha :as s]
            [clojure.walk :as walk]
            [clojure.java.io :as io]))

(def available-specs
  "All the available specs loaded by the system through the `setup!` function."
  (atom {}))

(defn- load-specs [filename]
  "If the file is not complaint with the usage of `:intent` and `:location`
a exception will be raised."
  (let [single-spec (->> filename
                         slurp
                         edn/read-string)]
    (->> (for [k (keys single-spec)]
           (if-not (s/valid? ::dsl/setup-file (get single-spec k))
             (throw (ex-info "Config file is out of specification format" (get single-spec k)))
             (hash-map k (get-in single-spec [k :location]))))
         (into {}))))

(defn setup! [files]
  "Perform the configuration of your pre-defined specs.

  :files   a list of java.io.File objects with the definitions
of your pre-defined specs. You should follow the specified documentation
about the content type of this files e.g. required keys `:intent` and
`:location`."
  (let [definitions (->> files
                         (filter #(.isFile %))
                         (map load-specs)
                         (into {}))]
    (if (empty? definitions)
      (throw (ex-info "No definitions found." {}))
      (swap! available-specs conj definitions))))


(defn- spec-name--or-clause [field-name spec-name]
  (let [fname (name field-name)
        sname (name spec-name)]
    (keyword (str fname "-" sname))))

(defn- spec-name--map-in-vector [nodes]
  (->> (keys nodes)
       (map name)
       (cstr/join "-")))

(def ^:private pre-targets (atom {}))

(defn- traverse-file-spec
  ([spec]
   (loop [k (keys spec)
          parser {}]
     (if (empty? k)
       parser
       (recur (rest k)
              (conj parser
                    (let [node (get spec (first k))
                          field-name (first k)]
                      (hash-map (if (:optional? node)
                                  (ds/opt field-name)
                                  field-name)
                                (cond
                                  ;; first level, already has a :spec keyword defining the data coercion
                                  (:spec node)
                                  (do
                                    (swap! pre-targets assoc field-name (:destination node))
                                    (if (:nullable? node)
                                      (ds/maybe (get @available-specs (:spec node)))
                                      (get @available-specs (:spec node))))

                                  ;; second level, we get a vector with a map containing :spec inside it.
                                  (and (vector? node) (:spec (first node)))
                                  (do
                                    (swap! pre-targets assoc field-name (:destination (first node)))
                                    (if (> (count node) 1)
                                      [(ds/or
                                        (->> (for [n node]
                                               (if (:spec n)
                                                 (hash-map (spec-name--or-clause field-name (:spec n)) (get @available-specs (:spec n)))
                                                 (hash-map (spec-name--or-clause field-name (spec-name--map-in-vector n)) (traverse-file-spec n))))
                                             (into {})))]
                                      [(get @available-specs (:spec (first node)))]))

                                  ;; third level, we get a vector with a new nested map inside, so call it again.
                                  (and (vector? node) (nil? (:spec (first node))))
                                  [(traverse-file-spec (first node))]

                                  ;; fourth level, we get a nested structure map without a :spec key to define the coercion
                                  :else (traverse-file-spec node))))))))))

(defn create-parser [map-spec]
  (if (empty? @available-specs)
    (throw (ex-info "You need to setup! your spec definitions first!" {}))
    (let [spec-name (:spec-name map-spec)
          spec-gen (traverse-file-spec (dissoc map-spec :spec-name))]
      (ds/spec {:name spec-name
                :spec spec-gen}))))

(defn- find-key [data ks matched-keys]
  (walk/walk (fn [[k v]]
               (cond
                 (and (contains? (set ks) k) (not (contains? (set @matched-keys) k))) (do (swap! matched-keys conj k) [k v])
                 (map? v) (find-key v ks matched-keys)
                 :else (into {} (map #(find-key % ks matched-keys) v)))) identity data))

(defn- find-keys [data ks]
  (let [matched-keys (atom [])]
    (find-key data ks matched-keys)))


(defn input-data->target-data
  "Convert an input json file to an output json file based on specifications.

  :data           INPUT data
  :data-spec      Map to describe the INPUT data and the desired TARGET location of each key through `:dest` parameters
  :target-shape   Map to describe the OUTPUT data format based on the `:dest` keyword placed in the `:spec-data`

  If the :target-fmt has a placeholder that was not
  fulfilled by the spec-data, it will return the placeholder
  instead of the desired data at that place"
  [data data-spec target-shape]
  (let [parser (create-parser data-spec)
        valid? (s/valid? parser data)]
    (if-not valid?
      (s/explain-data parser data)
      (-> @pre-targets
          (walk/postwalk-replace data)
          (find-keys (vals @pre-targets))
          (walk/postwalk-replace target-shape)))))


(comment

  (require '[clojure.java.io :as io])
  ;; TODO: Remover o atom de pre-targets do codigo e gerar essa variavel de outra forma
  ;; TODO: processo de post-walk está correto, porém vale a pena explorar edge cases com nomes duplicados e etc...

  (def spec-escrita-por-PO {:spec-name ::testando
                            :valor {:spec :teste
                                    }
                            :vals [{:spec :teste
                                    :dest :l}]
                            :valsb {:life {:casa {:spec :teste
                                                  :dest :x}}}})

  (def data-original {:valor 20
                      :vals [20 30 40 50]
                      :valsb {:life {:casa 99}}})

  (def target-fmt-conhecido {:new-val :v
                             })

  (setup! (-> (io/resource "specs")
              (io/file)
              (file-seq)))

  (def client-data {:zip ["101030-201", "987621-281"]
                    :rent 980.322
                    :university {:departments [{:zip {:address "University at Medium Inc.,"}}]}})

  (def client-spec {:spec-name :my-project.client/payload
                    :zip [{:spec :zipcode
                           :optional? false
                           :destination :zipcode
                           :nullable? false}]

                    :rent {:spec :money
                           :optional? false
                           :destination :value
                           :nullable? false}

                    :university {:departments [{:zip {:address {:spec :zipcode
                                                                :destination :university-address
                                                                :optional? false}}}]}})

  (def target-fmt {:my-internal-zipcode :zipcode
                   :my-internal-value   :value
                   :my-internal-address :university-address})

  (setup! (-> (io/resource "specs")
              (io/file)
              (file-seq)))

  (input-data->target-data client-data client-spec target-fmt)

  )
